cmd/kaf/consume.go

```diff
@@ -4,63 +4,113 @@ import (
 	"github.com/birdayz/kaf/pkg/proto"
 	"github.com/golang/protobuf/jsonpb"
-	prettyjson "github.com/hokaccha/go-prettyjson"
-	colorable "github.com/mattn/go-colorable"
+	"github.com/hokaccha/go-prettyjson"
+	"github.com/mattn/go-colorable"
 	"github.com/spf13/cobra"
 )
 
 var (
-	offsetFlag  string
-	raw         bool
-	follow      bool
-	schemaCache *avro.SchemaCache
-	keyfmt      *prettyjson.Formatter
+	partitionsWhitelistFlag []int
+	offsetFlags             []string
+	countFlags              []int
+	endFlag                 bool
+	raw                     bool
+	followFlag              bool
+	tailFlag                int
+	schemaCache             *avro.SchemaCache
+	keyfmt                  *prettyjson.Formatter
 
 	protoType string
 )
 
 func init() {
 	rootCmd.AddCommand(consumeCmd)
-	consumeCmd.Flags().StringVar(&offsetFlag, "offset", "oldest", "Offset to start consuming. Possible values: oldest, newest.")
-	consumeCmd.Flags().BoolVar(&raw, "raw", false, "Print raw output of messages, without key or prettified JSON")
-	consumeCmd.Flags().BoolVarP(&follow, "follow", "f", false, "Shorthand to start consuming with offset HEAD-1 on each partition. Overrides --offset flag")
-	consumeCmd.Flags().StringSliceVar(&protoFiles, "proto-include", []string{}, "Path to proto files")
-	consumeCmd.Flags().StringSliceVar(&protoExclude, "proto-exclude", []string{}, "Proto exclusions (path prefixes)")
-	consumeCmd.Flags().StringVar(&protoType, "proto-type", "", "Fully qualified name of the proto message type. Example: com.test.SampleMessage")
+	consumeCmd.Flags().StringSliceVarP(&offsetFlags, "offset", "o", []string{},
+		`Offset to start consuming. Pass a single value to apply to all partitions; pass a comma-delimited list to apply to specific partitions.
+Possible values: oldest, newest, -X, +X, X. (X is a number)
+oldest: Start reading at the oldest offset. This is the default.
+newest: Start reading at the newest offset.
+X: Start reading at offset X.
+-X: Start reading X before newest offset.
++X: Start reading X after oldest offset.`)
+	consumeCmd.Flags().IntSliceVarP(&countFlags, "count", "c", []int{}, "Number of messages to consume before exiting. Specify a single value for global limit, a comma-delimited list to limit specific partitions.")
+	consumeCmd.Flags().IntSliceVarP(&partitionsWhitelistFlag, "partitions", "p", []int{}, "Partitions to read from as a comma-delimited list; if unset, all partitions will be read.")
+	consumeCmd.Flags().BoolVar(&raw, "raw", false, "Print raw output of messages, without key or prettified JSON.")
+	consumeCmd.Flags().BoolVarP(&followFlag, "follow", "f", false, "Shorthand to start consuming with offset HEAD-1 on each partition. Overrides --offset flag.")
+	consumeCmd.Flags().IntVarP(&tailFlag, "tail", "t", 0, "Reads the last X messages by distributing X over all partitions. Overrides --offset flag.")
+	consumeCmd.Flags().BoolVarP(&endFlag, "end", "e", false, "Stop reading after reaching the end of the partition.")
```

Comment:
is this needed ? can't a user just set --follow to false ?

Reply:
Extending on https://github.com/birdayz/kaf/pull/67#discussion_r324420501, I didn't think I should mess with the default behavior of `consume` waiting for new messages to arrive when it reaches the end of a partition, and setting `--follow` to true already means "read from max offset - 1" so I needed a way to tell kaf to stop at the end of the partition(s). I guess we could maybe have 

- `> kaf` (no follow flag): start at beginning, keep waiting for messages
- `> kaf --follow`: start at max-1, keep waiting for messages
- `> kaf --follow=false`: start at beginning, stop at end of partition

and use https://stackoverflow.com/a/54747682/2611587 to check if the flag was provided, but that seems a bit confusing.